039-var-sources/proposal.md

```diff
@@ -0,0 +1,187 @@
+
+## "There can be only one"
+
+Only supporting a single credential manager really limits the possibilities of
+using credential managers for specialized use cases.
+
+A core tenent Concourse resources is that their content, i.e. version history
+and bits, should be addressable solely by the resource's configuration. That
+is, given a resource's `type:` and `source:`, the same version history will be
+returned on any Concourse installation, and can therefore be de-duped and
+shared across teams within an installation. This means not relying on cluster
+state for access control; resource types should entirely trust their `source:`.
+
+This is problematic for resources which make use of IAM roles associated to
+their `worker` EC2 instances in order to authenticate, because in this case the
+resource's `source:` does not actually include any credentials. As a result, we
+cannot safely enable [global resources][global-resources-opt-out]. by default
+because these resources would share version history without even vetting their
+credentials.
+
+A special credential manager could be implemented to acquire credentials via
+IAM roles on the `web` EC2 instance and then provide them to the `source:`
+configuration via `((vars))`. This way the `source:` configuration is the
+source of truth. This is discussed in [concourse/concourse#3023][issue-3023].
+
+However, as there can only be one credential manager configured at a time,
+using that single "slot" just for IAM roles is a bit of a waste compared to a
+full-fledged credential manager that can be used for many more things.
+
+# Proposal
+
+This proposal introduces a new kind of configuration: `var_sources`.
+
+This name "var source" is chosen to build on the existing terminology around
+`((vars))` and to directly relate them to one another. Calling them "var
+sources" instead of "credential managers" also allows them to be used for
+things that aren't necessarily credentials.
+
+`var_sources` may be specified at a pipeline-level, like so:
+
+```yaml
+var_sources:
+- name: vault
+  type: vault
+  config:
+    uri: https://vault.example.com
+    # ... vault-specific config including auth/etc ...
+
+resources: # ...
+
+jobs: # ...
+```
+
+Each var source has a `name` which must be a [valid
+identifier][valid-identifier-rfc]. This is used to explicitly reference the
+source from `((vars))` syntax so that there is no ambiguity. See
+[`VAR_SOURCE_NAME`](#VAR_SOURCE_NAME).
+
+A var source's `type` specifies one of the supported credential managers, e.g.
+`vault`, `credhub`, `kubernetes`. is responsible for interpreting
+`config`.
+
+A var source's `config` is a "black box" to Concourse and is passed verbatim to
+the credential manager. This configuration should include any credentials
+necessary for authenticating with the credential manager.
+
+A var source's `config` may use `((vars))` to obtain its own credentials,
+either by using a different var source or by using the system-level credential
```

Comment:
I think a `var_source`'s `config` should only get var from the system credential manager.

Reply:
I don't feel *too* strongly about this, but I decided to allow it so that it's consistent with how `resource_types` works - resource types can use *other* resource types to fetch their image, so it seemed only fair to allow the same of var sources. Do you think this would be particularly challenging to support?

Reply:
I feel challenging of implementation. It should be done in `db.pipeline.Varaibles()`, a small change will support that. My concern is that would lead some potential complexity, for example, cycle reference:

```yml
var_sources:
- name: vs1
  type: vault
  config: 
      token: ((vs2:mytoken))  
- name: vs2
  type: vault
  config: 
      token: ((vs1:mytoken))  
```

If we support get var_source's var from other var_source, then we will have to detect such cycle reference problem, maybe other issues also. But `resource_type` doesn't have such problem.

So I think, at least, we start with not supporting that. If users raise such request in future, we may consider to add such support.

Reply:
`resource_types` actually does have the same problem:

```yaml
resource_types:
- name: foo
  type: bar
  source: # ...

- name: bar
  type: foo
  source: # ...
```

The implementation is pretty simple: if resource type `bar` is being used, then `bar` is ignored when recursively resolving its type `foo`. This trick also allows for e.g. the `registry-image` base resource type to be overridden, using the base resource type to fetch its replacement image:

```yaml
resource_types:
- name: registry-image # same
  type: registry-image # same
  source: # replacement
```

I think the same could probably be done here; instead of going into an infinite loop, an `UnknownVarSourceError` could be returned, and we could print a nice error if the 'unknown source' name is the same as the one we're currently resolving.

In any case, I'm OK with waiting for users to ask for it, since it'd be easy enough to implement later. It just felt like something worth specifying one way or the other, and I can't think of any drawbacks I figured I'd just allow it.

How about I move this under 'Open Questions' and we can collect more feedback on it?

Reply:
> How about I move this under 'Open Questions' and we can collect more feedback on it?

That's a good idea.

Reply:
Moved it to 'open questions'. I have to say as I wrote it out it kind of feels sensible to allow it, primarily for the 'bootstrapping' use case, but we can talk it out there.